### PR TITLE
sys-libs/libsemanage, sys-apps/policycoreutils: Make python dependency optional

### PR DESCRIPTION
**Pull request by krnowak**
This PR tries to make python dependency optional in aforementioned packages. The reason is that in Flatcar we don't provide python interpreter in the production images, so for a long time we were carrying modified versions of the Gentoo ebuilds that were stripping all the python deps for our needs. Originally, we were just doing this - removing the deps, removing all the scripting that was dealing with python. With this PR we are trying to achieve the same goal, but in an upstreamable way.

Please see the commit messages, where the changes made are described in a bit greater detail.

Thanks!

CC: @tormath1

### CODE REVIEW

**Review by 0xC0ncord — COMMENTED** [review]
From a quick glance it looks good! I haven't tested it yet though.

My first thought is I wonder how many users out there may have deliberately disabled `USE=python` systemwide without really knowing any better, and whether or not this would cause unexpected breakage.

**Comment by krnowak** [comment]
> From a quick glance it looks good! I haven't tested it yet though.

Thanks! Good to know.

> My first thought is I wonder how many users out there may have deliberately disabled `USE=python` systemwide without really knowing any better, and whether or not this would cause unexpected breakage.

Yeah, I was thinking about it too, that's why I made `python` USE enabled-by-default. Obviously won't do the trick in your scenario, but I'm not sure if this isn't a self-inflicted wound in this case.

So I dunno, maybe a temporary use.force entry for those two packages to be dropped after X days/weeks?

**Comment by 0xC0ncord** [comment]
> Yeah, I was thinking about it too, that's why I made `python` USE enabled-by-default. Obviously won't do the trick in your scenario, but I'm not sure if this isn't a self-inflicted wound in this case.
> 
> So I dunno, maybe a temporary use.force entry for those two packages to be dropped after X days/weeks?

My concern is mostly whether or not if Portage depends on the Python functionality or serious problems will occur. I can see that Portage pulls in the Python bits but I don't know enough about Portage internals (at the time of writing this) if these code paths will end up calling into `libselinux` directly. If that's the case then we can just force-enable this flag for the SELinux profiles in Gentoo, otherwise I'm of the opinion that this is more of a documentation issue than a technical one.

What do you think @perfinion ?

**Comment by krnowak** [comment]
> > Yeah, I was thinking about it too, that's why I made `python` USE enabled-by-default. Obviously won't do the trick in your scenario, but I'm not sure if this isn't a self-inflicted wound in this case.
> > So I dunno, maybe a temporary use.force entry for those two packages to be dropped after X days/weeks?
> 
> My concern is mostly whether or not if Portage depends on the Python functionality or serious problems will occur. I can see that Portage pulls in the Python bits but I don't know enough about Portage internals (at the time of writing this) if these code paths will end up calling into `libselinux` directly. If that's the case then we can just force-enable this flag for the SELinux profiles in Gentoo, otherwise I'm of the opinion that this is more of a documentation issue than a technical one.

Gotcha.

One interesting bit that I have found so far is that there is `preinst_selinux_labels` function in portage ([`bin/misc-functions.sh`](https://github.com/gentoo/portage/blob/portage-3.0.63/bin/misc-functions.sh#L473-L498)) that checks for `/usr/sbin/selinuxconfig`, which is provided by policycoreutils-extra tarball. That would mean that for this function to do its work, policycoreutils would need to be installed with `USE=-vanilla` (which, with my changes, also pulls `USE=python`). But `selinuxconfig` is a compiled binary, then maybe I could change the ebuild in a way that `vanilla` is independent from `python`. Basically `USE=-vanilla` would pull the policycoreutils-extra tarball and install the binaries. If `USE=python` is also enabled, then it would also install the python scripts from the extra tarball.

So I guess that the `selinux` profile should force `USE=-vanilla` on `sys-apps/policycoreutils` then.

**Comment by krnowak** [comment]
Hm. `rlpkg` script seems to be used by `sec-policy/selinux-base-policy` in `pkg_postinst` and by `selinux-policy-2.eclass` in its `pkg_postinst` implementation. But only when not cross-compiling. It seems that `selinux` profile would need `USE-vanilla python` for `sys-apps/policycoreutils`.

I was looking into it a bit deeper, because we have Flatcar SDK, which is built using catalyst, so it's not doing any cross-compilation, thus should be using `rlpkg`. But we have modified policycoreutils ebuild to never install `rlpkg` and I have never seen any error about `rlpkg` not being found. Looking at the `pkg_postinst` function it has the following code:

```bash
local PKGSET="";
if [[ -x /usr/bin/qdepends ]] ; then
	PKGSET=$(/usr/bin/qdepends -Cq -r -Q ${CATEGORY}/${PN} | grep -v 'sec-policy/selinux-');
elif [[ -x /usr/bin/equery ]] ; then
	PKGSET=$(/usr/bin/equery -Cq depends ${CATEGORY}/${PN} | grep -v 'sec-policy/selinux-');
fi
if [[ -n "${PKGSET}" ]] ; then
	rlpkg ${PKGSET};
fi
```

Using `qdepends` in Flatcar SDK for `sec-policy/selinux-base-policy` returns:

```
sec-policy/selinux-unconfined-2.20231002-r2: >=sec-policy/selinux-base-policy-2.20231002-r2
sys-apps/systemd-252.11-r2: sec-policy/selinux-base-policy[systemd]
```

And using `equery`:

```
sec-policy/selinux-unconfined-2.20231002-r2
sys-apps/systemd-252.11-r2
```

So `grep -v sec-policy/selinux-` is going to filter out the whole output from `qdepends`, while `equery`'s output will contain `sys-apps/systemd`. In Flatcar SDK we have `/usr/bin/qdepends`, so `equery` won't get used and `rlpkg` won't be invoked.

Couple of takeaways:

- Currently `selinux` profile would need to have `USE=-vanilla python`.
- There seem to be a bug in filtering output from `qdepends`. Looks like it needs a smarter grep query and then each line needs to be truncated at `:`.
- The `rlpkg` script usage could probably be replaced with `equery -C files sys-apps/systemd 2>/dev/null | sed -e 's/ ->.*//' | /sbin/restorecon -f -`.

Please let me know what you think.

**Comment by krnowak** [comment]
> Couple of takeaways:
> 
> * Currently `selinux` profile would need to have `USE=-vanilla python`.
> * There seem to be a bug in filtering output from `qdepends`. Looks like it needs a smarter grep query and then each line needs to be truncated at `:`.
> * The `rlpkg` script usage could probably be replaced with `equery -C files sys-apps/systemd 2>/dev/null | sed -e 's/ ->.*//' | /sbin/restorecon -f -`.
> 
> Please let me know what you think.

I made a stab at addressing the last two points from the takeaways list (+ code deduplication), please see the PR: https://github.com/gentoo/gentoo/pull/35747

File: sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	S1="${WORKDIR}/${MY_P}"
+	S2="${WORKDIR}/policycoreutils-extra"
+	S="${S1}"
+fi
+
+LICENSE="GPL-2"
+SLOT="0"
+IUSE="audit pam +python split-usr vanilla"
+REQUIRED_USE="
+	!vanilla? ( python ${PYTHON_REQUIRED_USE} )
```
Comment (perfinion) [inline sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild:37]: hmm, after reading this a bit im not sure I entirely follow your intention for `USE=vanilla vs USE=python`

1. I assume you actually just dont want `rlpkg` ?
2. Do you care about anything else in policycoreutils-extra at all? either for or against?
3. If i changed the policycoreutils-extra Makefiles around to just skip rlpkg but always install all the rest (ie the couple C utils and selocal which is a bash script), is that better?
4. Long term I kind of want to move `rlpkg` into policycoreutils proper and just get rid of everything else. (`selocal` should stay, not sure if it should be upstream or be the only thing in extras)

File: sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	!python? (
+		>=sys-libs/libselinux-${PV}:=
+		>=sys-libs/libsemanage-${PV}:=
+		audit? ( >=sys-process/audit-1.5.1 )
+	)
+	>=sys-libs/libsepol-${PV}:=
+	sys-libs/libcap-ng:=
+	pam? ( sys-libs/pam:= )
+	!vanilla? (
+		>=app-admin/setools-4.2.0[${PYTHON_USEDEP}]
```
Comment (perfinion) [inline sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild:56]: I'm not following why setools isnt just in the `python?()` section?
The `REQUIRED_USE` means if vanilla is not set, then python is required?

File: sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+src_prepare() {
+	S="${S1}"
+	cd "${S}" || die "Failed to switch to ${S}"
+	if [[ ${PV} != 9999 ]] ; then
+		# If needed for live ebuilds please use /etc/portage/patches
+		eapply "${FILESDIR}/policycoreutils-3.1-0001-newrole-not-suid.patch"
+	fi
+
+	if ! use vanilla; then
+		# rlpkg is more useful than fixfiles
```
Comment (perfinion) [inline sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild:88]: oh, do you actually want `fixfiles` instead of rlpkg? or was this just cuz rlpkg is removed for vanilla? Cuz last i looked (many many years ago to be fair), fixfiles was really redhat-specific and didnt make a ton of sense on gentoo. I'll need to look again but it should either be unconditionally added or unconditionally never there. Its such a tiny binary that (assuming it works fine on gentoo) there is no harm in always having it.

File: sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+			|| die "fixfiles sed 1 failed"
+		sed -i -e '/fixfiles/d' "${S}/scripts/Makefile" \
+			|| die "fixfiles sed 2 failed"
+	fi
+
+	eapply_user
+
+	sed -i 's/-Werror//g' "${S1}"/*/Makefile || die "Failed to remove Werror"
+
+	if ! use vanilla; then
```
Comment (perfinion) [inline sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild:99]: this needs to be `if ! use python`

File: sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+"
+
+DEPEND="
+	python? (
+		>=sys-libs/libselinux-${PV}:=[python,${PYTHON_USEDEP}]
+		>=sys-libs/libsemanage-${PV}:=[python(+),${PYTHON_USEDEP}]
+		audit? ( >=sys-process/audit-1.5.1[python,${PYTHON_USEDEP}] )
+		${PYTHON_DEPS}
+	)
+	!python? (
```
Comment (perfinion) [inline sys-apps/policycoreutils/policycoreutils-3.6-r1.ebuild:47]: I need need to look up some things again, pretty sure there is an easier way to do this tho.

File: sys-libs/libsemanage/files/libsemanage-extra-config.patch [hunk 6]
```diff
@@ -0,0 +1,30 @@
+diff -r -u libsemanage-3.6/src/semanage.conf libsemanage-3.6-patched/src/semanage.conf
+--- libsemanage-3.6/src/semanage.conf	2023-12-13 15:46:22.000000000 +0100
++++ libsemanage-3.6-patched/src/semanage.conf	2024-02-21 14:33:06.055611733 +0100
```
Comment (perfinion) [inline sys-libs/libsemanage/files/libsemanage-extra-config.patch:3]: oh man, i really just need to upstream this part haha.

**Review by perfinion — CHANGES_REQUESTED** [review]
Just a first pass, i need to look closer at a few things, no need for a new revision of this yet.

**Comment by perfinion** [comment]
> > > Yeah, I was thinking about it too, that's why I made `python` USE enabled-by-default. Obviously won't do the trick in your scenario, but I'm not sure if this isn't a self-inflicted wound in this case.
> > > So I dunno, maybe a temporary use.force entry for those two packages to be dropped after X days/weeks?

heh, definitely not days. I'd need to `package.use.force` for like minimum a year (selinux is in the `@system` set, it needs to work for extremely infrequent updates.). The better option is to add `[python]` to the dep in portage then wait til all old versions of portage are dropped then we can drop the package.use.force.

> > My concern is mostly whether or not if Portage depends on the Python functionality or serious problems will occur. I can see that Portage pulls in the Python bits but I don't know enough about Portage internals (at the time of writing this) if these code paths will end up calling into `libselinux` directly. If that's the case then we can just force-enable this flag for the SELinux profiles in Gentoo, otherwise I'm of the opinion that this is more of a documentation issue than a technical one.
> 
> Gotcha.
> 
> One interesting bit that I have found so far is that there is `preinst_selinux_labels` function in portage ([`bin/misc-functions.sh`](https://github.com/gentoo/portage/blob/portage-3.0.63/bin/misc-functions.sh#L473-L498)) that checks for `/usr/sbin/selinuxconfig`, which is provided by policycoreutils-extra tarball. That would mean that for this function to do its work, policycoreutils would need to be installed with `USE=-vanilla` (which, with my changes, also pulls `USE=python`). But `selinuxconfig` is a compiled binary, then maybe I could change the ebuild in a way that `vanilla` is independent from `python`. Basically `USE=-vanilla` would pull the policycoreutils-extra tarball and install the binaries. If `USE=python` is also enabled, then it would also install the python scripts from the extra tarball.

`selinuxconfig` is extremely legacy. The path forward for this all is most likely to have `USE=-python` remove `rlpkg` a keep the rest of extras, then there is no need for USE=vanilla. Then I'll also fix portage to use a binary from upstream, then eventually just drop `selinuxconfig` from extras (or just all of extras).

> So I guess that the `selinux` profile should force `USE=-vanilla` on `sys-apps/policycoreutils` then.

I assume in flatcar you have your own profiles entirely? or are you layering on top of the gentoo ones? Even if you are layering on top then it's pretty easy for you to unforce it.

**Comment by perfinion** [comment]
> Hm. `rlpkg` script seems to be used by `sec-policy/selinux-base-policy` in `pkg_postinst` and by `selinux-policy-2.eclass` in its `pkg_postinst` implementation. But only when not cross-compiling. It seems that `selinux` profile would need `USE-vanilla python` for `sys-apps/policycoreutils`.
> 
> I was looking into it a bit deeper, because we have Flatcar SDK, which is built using catalyst, so it's not doing any cross-compilation, thus should be using `rlpkg`. But we have modified policycoreutils ebuild to never install `rlpkg` and I have never seen any error about `rlpkg` not being found. Looking at the `pkg_postinst` function it has the following code:

Heh, so there is some history here, finally found the CVS repo, I had forgotten how annoying CVS was. here are some links:
https://gitweb.gentoo.org/repo/gentoo/historical.git/commit/sec-policy/selinux-base-policy?id=124abe0befb75958dbcb310f8857b3c9c11bfc2d
https://gitweb.gentoo.org/repo/gentoo/historical.git/commit/sec-policy/selinux-base-policy?id=556c9362c16e79f47c2825efe3637846986ebd29
"Relabel depending packages so we no longer need DEPEND calls for pure policy dependencies in SELinux"
I can't seem to find anything in bugzilla.

The problem if i recall it was if the policy package selinux-foo was installed after foo. If foo has RDEP=selinux-foo and you have none of them installed, then portage will install selinux-foo first, then build foo. in the postinst phase right before merging from `$D` into /, portage will setfiles all of the files properly then move into /. The issue is if you had `foo` installed but there was no `selinux-foo` (either because RDEP was missing previously, or because the next policy release added new fcontexts. Then the fcontexts on the files from foo on disk were incorrect. Next time foo is updated or reinstalled then portage will fix everything up no problem, but that might take a long time or its just a waste of time to rebuild for some fcontexts. With this part when a policy package is installed, it will find any pre-existing packages that depend on the policy and will force relabel them.

There is no need to relabel the files from selinux policy packages, because the fcontexts covering where those are installed are never optional, they're either in selinux-base or base-policy. There is no harm to doing it, its just a waste. 

One option was to always have selinux-foo in DEPEND vs RDEPEND, but that isnt actually correct either. There has been talk about an IDEPENDS (install-dep), if that is in a future EAPI then that is the right way. Portage behaviour is that RDEPEND is installed before which is close enough. 

Another option was to force the policy packages to depend on gentoolkit or one of those, but we instead decided its pretty likely that at least one of the tools exists, and in the unlikely case none of them exists, its not really a big deal since its such an edge-case.

TLDR, you almost definitely will not care at all because you're building with catalyst and installing everything from scratch and the order is correct, or you just relabel everything at the end.
